tests/simplex/step-12c.cc

```diff
@@ -0,0 +1,1158 @@
+
+  setup_system();
+  assemble_system2();
+
+  deallog << "   Number of degrees of freedom: " << dof_handler.n_dofs()
+          << std::endl;
+  solve(solution1);
+
+  output_results(0);
+}
+
+int
+main()
+{
+  try
+    {
+      initlog();
+      deallog.get_file_stream().precision(2);
+
+      // triangle
+      for (unsigned int i = 1; i <= 2; ++i)
+        {
+          ScratchData<2> scratch_data;
+
+          scratch_data.mapping =
+            hp::MappingCollection<2>(MappingFE<2>(Simplex::FE_P<2>(1)));
+          scratch_data.fe = hp::FECollection<2>(Simplex::FE_DGP<2>(i));
+          scratch_data.quadrature =
+            hp::QCollection<2>(Simplex::QGauss<2>(i + 1));
+          scratch_data.face_quadrature = std::vector<hp::QCollection<1>>{
+            hp::QCollection<1>(Simplex::QGauss<1>(i + 1))};
+          scratch_data.mesh_generator =
+            [](Triangulation<2> &triangulation) -> void {
+            GridGenerator::subdivided_hyper_cube_with_simplices(triangulation,
+                                                                16);
+          };
+
+          DGMethod<2> dgmethod(scratch_data);
+          dgmethod.run();
+        }
+
+      // quadrilateral
+      for (unsigned int i = 1; i <= 2; ++i)
+        {
+          ScratchData<2> scratch_data;
+
+          scratch_data.mapping =
+            hp::MappingCollection<2>(MappingQGeneric<2>(1));
+          scratch_data.fe              = hp::FECollection<2>(FE_DGQ<2>(i));
+          scratch_data.quadrature      = hp::QCollection<2>(QGauss<2>(i + 1));
+          scratch_data.face_quadrature = std::vector<hp::QCollection<1>>{
+            hp::QCollection<1>(QGauss<1>(i + 1))};
+          scratch_data.mesh_generator =
+            [](Triangulation<2> &triangulation) -> void {
+            GridGenerator::hyper_cube(triangulation);
+            triangulation.refine_global(3);
+          };
+
+          DGMethod<2> dgmethod(scratch_data);
+          dgmethod.run();
+        }
+
+      // tetrahedron
+      for (unsigned int i = 1; i <= 2; ++i)
+        {
+          ScratchData<3> scratch_data;
+
+          scratch_data.mapping =
+            hp::MappingCollection<3>(MappingFE<3>(Simplex::FE_P<3>(1)));
+          scratch_data.fe = hp::FECollection<3>(Simplex::FE_DGP<3>(i));
+          scratch_data.quadrature =
+            hp::QCollection<3>(Simplex::QGauss<3>(i + 1));
+          scratch_data.face_quadrature = std::vector<hp::QCollection<2>>{
+            hp::QCollection<2>(Simplex::QGauss<2>(i + 1))};
+          scratch_data.mesh_generator =
+            [](Triangulation<3> &triangulation) -> void {
+            GridGenerator::subdivided_hyper_cube_with_simplices(triangulation,
+                                                                8);
+          };
+
+          DGMethod<3> dgmethod(scratch_data);
+          dgmethod.run();
+        }
+
+      // pyramid
+      for (unsigned int i = 1; i <= 1; ++i)
+        {
+          continue;
```

Comment:
I can enable the pyramid and the wedge test once #11401 is merged.

Reply:
Matrix-free and matrix-based results for linear and quadratic discontinuous simplex. hypercube, pyramid, and wedge elements coincide.